Alicerce/Sources/Utils/ServiceLocator.swift

```diff
@@ -8,104 +8,155 @@
-    }
-    
-    func register<Service>(name serviceName: String? = nil, _ lazyInit: @escaping LazyServiceClosure<Service>) throws {
+
+    @discardableResult
+    func register<Service>(name serviceName: ServiceName? = nil, service: Service) throws -> ServiceName {
         let name = buildName(for: Service.self, serviceName)
-        
-        try checkForDuplicateService(with: name)
-        
-        lazyServiceInits[name] = lazyInit
+
+        try synchronized {
+            try validate(serviceName: name)
+            services[name] = .normal(service)
+        }
+
+        return name
     }
-    
-    func unregister<Service>(_: Service, name serviceName: String? = nil) throws {
+
+    @discardableResult
+    func register<Service>(name serviceName: ServiceName? = nil,
+                           _ lazyInit: @escaping LazyInit<Service>) throws -> ServiceName {
         let name = buildName(for: Service.self, serviceName)
-        
-        guard let anyService = services[name] ?? lazyServiceInits[name]?() else {
-            throw ServiceLocatorError.inexistentService
+
+        try synchronized {
+            try validate(serviceName: name)
+            services[name] = .lazy(lazyInit)
         }
-        
-        guard let _ = anyService as? Service else {
-            throw ServiceLocatorError.serviceTypeMismatch(expected: "\(Service.self)", found: "\(type(of: anyService))")
+
+        return name
+    }
+
+    func unregister<Service>(_ type: Service.Type, name serviceName: ServiceName? = nil) throws {
+        let name = buildName(for: type, serviceName)
+
+        try synchronized {
+            guard let service = services[name] else { throw ServiceLocatorError.inexistentService }
+
+            try validate(serviceType: type, service: service)
+            services[name] = nil
         }
-        
-        services[name] = nil
-        lazyServiceInits[name] = nil
     }
-    
+
     func unregisterAll() {
-        services.removeAll()
-        lazyServiceInits.removeAll()
-    }
-    
-    func get<Service>(name serviceName: String? = nil) throws -> Service {
-        return try locate(serviceName)
+        synchronized {
+            services.removeAll()
+        }
     }
-    
-    // MARK: - Private Methods
-    
-    private func locate<Service>(_ serviceName: String? = nil) throws -> Service {
+
+    func get<Service>(name serviceName: ServiceName? = nil) throws -> Service {
         let name = buildName(for: Service.self, serviceName)
-        if let lazyServiceClosure = lazyServiceInits[name] {
-            let service: Service = try checkIfServiceValid(lazyServiceClosure())
-            
-            // Remove closure since the service is now validated
-            lazyServiceInits[name] = nil
-            
-            // This should *never* fail unless we have some name collision bug, so 💥 immediately
-            try! register(name: name, service: service)
+        var service: Service?
```

Comment:
We could use ! directly if we know that the variable will be populated. Or do you want to fail if it don't?

This should be a validate responsibility, i think.